File: proyecto_java_estatuto/src/main/java/co/edu/udistrital/estatuto/modelo/UnidadAcademica.java
package co.edu.udistrital.estatuto.modelo;

import java.time.LocalDate;
import java.util.Objects;

/**
 * Clase abstracta que representa una unidad académico-administrativa
 * de la Universidad Distrital Francisco José de Caldas.
 *
 * <p>Conforme al Acuerdo 004 de 2025 (Estatuto General), toda unidad
 * académica posee un nombre, un código institucional, una fecha de
 * creación oficial y un director designado por el órgano competente.</p>
 *
 * <p>Las subclases concretas son: {@link Facultad}, {@link Escuela},
 * {@link Centro} e {@link Instituto}.</p>
 *
 * @author Roberto Albeiro Pava-Díaz
 * @version 1.0
 * @see Facultad
 * @see Escuela
 * @see Centro
 * @see Instituto
 */
public abstract class UnidadAcademica {

    /** Nombre oficial de la unidad académica. */
    private String nombre;

    /** Código institucional único asignado por la Rectoría. */
    private String codigo;

    /** Fecha de creación o reconocimiento oficial de la unidad. */
    private LocalDate fechaCreacion;

    /** Nombre completo del director o directora vigente. */
    private String director;

    /**
     * Construye una unidad académica con los datos básicos.
     *
     * @param nombre        nombre oficial de la unidad (no nulo, no vacío)
     * @param codigo        código institucional (no nulo, no vacío)
     * @param fechaCreacion fecha de creación oficial
     * @param director      nombre del director vigente
     * @throws IllegalArgumentException si el código o el nombre son nulos/vacíos
     * @throws NullPointerException     si el nombre es nulo
     */
    protected UnidadAcademica(String nombre, String codigo,
                               LocalDate fechaCreacion, String director) {
        if (codigo == null || codigo.isBlank()) {
            throw new IllegalArgumentException(
                "El código de la unidad académica no puede ser nulo ni vacío.");
        }
        this.nombre = Objects.requireNonNull(nombre, "El nombre no puede ser nulo.");
        this.codigo = codigo.trim().toUpperCase();
        this.fechaCreacion = fechaCreacion;
        this.director = director;
    }

    /**
     * Retorna el tipo de unidad académica (Facultad, Escuela, Centro, etc.).
     * Cada subclase debe implementar este método con su denominación específica.
     *
     * @return cadena descriptiva del tipo de unidad
     */
    public abstract String getTipo();

    // ── Getters y Setters ──────────────────────────────────────────────────

    /**
     * Retorna el nombre oficial de la unidad.
     * @return nombre de la unidad
     */
    public String getNombre() {
        return nombre;
    }

    /**
     * Actualiza el nombre oficial de la unidad.
     * @param nombre nuevo nombre (no nulo, no vacío)
     * @throws IllegalArgumentException si el nombre es nulo o vacío
     */
    public void setNombre(String nombre) {
        if (nombre == null || nombre.isBlank()) {
            throw new IllegalArgumentException("El nombre no puede ser vacío.");
        }
        this.nombre = nombre.trim();
    }

    /**
     * Retorna el código institucional.
     * @return código en mayúsculas
     */
    public String getCodigo() {
        return codigo;
    }

    /**
     * Actualiza el código institucional. Se normaliza a mayúsculas.
     * @param codigo nuevo código (no nulo, no vacío)
     * @throws IllegalArgumentException si el código es nulo o vacío
     */
    public void setCodigo(String codigo) {
        if (codigo == null || codigo.isBlank()) {
            throw new IllegalArgumentException("El código no puede ser vacío.");
        }
        this.codigo = codigo.trim().toUpperCase();
    }

    /**
     * Retorna la fecha de creación oficial.
     * @return fecha de creación
     */
    public LocalDate getFechaCreacion() {
        return fechaCreacion;
    }

    /**
     * Actualiza la fecha de creación.
     * @param fechaCreacion nueva fecha de creación
     */
    public void setFechaCreacion(LocalDate fechaCreacion) {
        this.fechaCreacion = fechaCreacion;
    }

    /**
     * Retorna el nombre del director vigente.
     * @return nombre del director
     */
    public String getDirector() {
        return director;
    }

    /**
     * Actualiza el nombre del director.
     * @param director nombre del nuevo director
     */
    public void setDirector(String director) {
        this.director = director;
    }

    // ── equals, hashCode, toString ─────────────────────────────────────────

    /**
     * Dos unidades académicas son iguales si tienen el mismo código.
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof UnidadAcademica u)) return false;
        return codigo.equals(u.codigo);
    }

    @Override
    public int hashCode() {
        return Objects.hash(codigo);
    }

    @Override
    public String toString() {
        return String.format("%s[codigo=%s, nombre=%s, director=%s]",
                             getTipo(), codigo, nombre, director);
    }
}

File: proyecto_rust_estatuto/src/modelo/unidad_academica.rs
//! # Trait `UnidadAcademica` y struct `UnidadBase`
//!
//! Define la abstracción base para todas las unidades académico-administrativas
//! de la Universidad Distrital Francisco José de Caldas, conforme al
//! **Artículo 8 del Acuerdo 004 de 2025**.
//!
//! ## Equivalencia Java → Rust
//! La clase abstracta `UnidadAcademica` de Java se convierte en:
//! - `trait UnidadAcademica`: define la interfaz polimórfica.
//! - `struct UnidadBase`: contiene los campos comunes (composición).
//!
//! Las estructuras concretas (`Facultad`, `Escuela`, etc.) contienen un campo
//! `base: UnidadBase` y deben implementar el trait `UnidadAcademica`.

use crate::excepcion::errores::EstatutoResult;
use crate::modelo::validador;
use chrono::NaiveDate;
use serde::{Deserialize, Serialize};
use std::fmt;

// ============================================================
//  UnidadBase — campos compartidos por todas las unidades
// ============================================================

/// Estructura con los campos comunes a todas las unidades académicas.
///
/// Equivale a los campos de la clase abstracta `UnidadAcademica` en Java.
/// Se usa mediante composición: cada unidad concreta tiene un campo `base`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnidadBase {
    /// Nombre oficial de la unidad académica.
    pub nombre: String,

    /// Código único identificador (normalizado en mayúsculas, sin espacios).
    ///
    /// Conforme al Acuerdo 004 de 2025, cada unidad tiene un código institucional
    /// único que sirve como clave de negocio.
    pub codigo: String,

    /// Fecha en que la unidad fue creada o reconocida formalmente.
    pub fecha_creacion: NaiveDate,

    /// Nombre del director o decano actual de la unidad.
    pub director: String,
}

impl UnidadBase {
    /// Crea una nueva `UnidadBase` con validación del código.
    ///
    /// El código es normalizado (mayúsculas, sin espacios extremos) antes de
    /// almacenarse, lo que equivale al comportamiento del constructor Java.
    ///
    /// # Errores
    /// Retorna [`EstatutoError::Validacion`] si:
    /// - `codigo` está vacío o solo contiene espacios.
    /// - `nombre` está vacío.
    ///
    /// # Ejemplo
    /// ```
    /// use chrono::NaiveDate;
    /// use estatuto_ud::modelo::unidad_academica::UnidadBase;
    ///
    /// let fecha = NaiveDate::from_ymd_opt(2025, 1, 1).unwrap();
    /// let base = UnidadBase::nuevo("Facultad de Ingeniería", "FAC-ING", fecha, "Dr. Pérez").unwrap();
    /// assert_eq!(base.codigo, "FAC-ING");
    /// ```
    pub fn nuevo(
        nombre: &str,
        codigo: &str,
        fecha_creacion: NaiveDate,
        director: &str,
    ) -> EstatutoResult<Self> {
        validador::no_vacio(nombre, "nombre")?;
        validador::no_vacio(codigo, "codigo")?;
        validador::no_vacio(director, "director")?;

        Ok(UnidadBase {
            nombre: nombre.trim().to_string(),
            codigo: validador::normalizar_codigo(codigo),
            fecha_creacion,
            director: director.trim().to_string(),
        })
    }
}

impl fmt::Display for UnidadBase {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "[{}] {} — Director: {} (desde {})",
            self.codigo, self.nombre, self.director, self.fecha_creacion
        )
    }
}

// ============================================================
//  Trait UnidadAcademica — interfaz polimórfica
// ============================================================

/// Trait que define la interfaz de las unidades académico-administrativas.
///
/// Equivale a la clase abstracta `UnidadAcademica` de Java (con sus métodos
/// abstractos `getTipo()` y `getNombre()`).
///
/// ## Implementaciones
/// - [`crate::modelo::facultad::Facultad`]
/// - [`crate::modelo::escuela::Escuela`]
/// - [`crate::modelo::centro::Centro`]
/// - [`crate::modelo::instituto::Instituto`]
pub trait UnidadAcademica: fmt::Display {
    /// Retorna el tipo de unidad como cadena descriptiva.
    ///
    /// Equivale al método abstracto `getTipo()` de Java.
    fn tipo(&self) -> &str;

    /// Retorna una referencia inmutable al `UnidadBase` subyacente.
    fn base(&self) -> &UnidadBase;

    /// Retorna una referencia mutable al `UnidadBase` subyacente.
    fn base_mut(&mut self) -> &mut UnidadBase;

    /// Retorna el código único de la unidad (delegado a `UnidadBase`).
    fn codigo(&self) -> &str {
        &self.base().codigo
    }

    /// Retorna el nombre de la unidad (delegado a `UnidadBase`).
    fn nombre(&self) -> &str {
        &self.base().nombre
    }
}

// ============================================================
//  Pruebas unitarias
// ============================================================

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::NaiveDate;

    fn fecha_prueba() -> NaiveDate {
        NaiveDate::from_ymd_opt(2025, 3, 1).unwrap()
    }

    #[test]
    fn unidad_base_crea_correctamente() {
        let base = UnidadBase::nuevo("Facultad de Ingeniería", "FAC-ING", fecha_prueba(), "Dr. Pérez");
        assert!(base.is_ok());
        let base = base.unwrap();
        assert_eq!(base.codigo, "FAC-ING");
        assert_eq!(base.nombre, "Facultad de Ingeniería");
    }

    #[test]
    fn unidad_base_normaliza_codigo() {
        let base = UnidadBase::nuevo("Escuela", "  esc-01  ", fecha_prueba(), "Dir").unwrap();
        assert_eq!(base.codigo, "ESC-01");
    }

    #[test]
    fn unidad_base_rechaza_codigo_vacio() {
        let resultado = UnidadBase::nuevo("Facultad", "", fecha_prueba(), "Dir");
        assert!(resultado.is_err());
    }

    #[test]
    fn unidad_base_rechaza_nombre_vacio() {
        let resultado = UnidadBase::nuevo("", "FAC-001", fecha_prueba(), "Dir");
        assert!(resultado.is_err());
    }

    #[test]
    fn unidad_base_display() {
        let base = UnidadBase::nuevo("Instituto", "INST-01", fecha_prueba(), "Dr. Gómez").unwrap();
        let texto = base.to_string();
        assert!(texto.contains("INST-01"));
        assert!(texto.contains("Instituto"));
    }

    #[test]
    fn unidad_base_serializa() {
        let base = UnidadBase::nuevo("Escuela", "ESC-01", fecha_prueba(), "Dir").unwrap();
        let json = serde_json::to_string(&base).unwrap();
        let back: UnidadBase = serde_json::from_str(&json).unwrap();
        assert_eq!(back.codigo, "ESC-01");
    }
}
